File: src/main/java/core/procedures/cons/ConsProc.java
package core.procedures.cons;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMCons;

public final class ConsProc extends AFn {

  public ConsProc() {
    super(new FnArgsBuilder().min(2).max(2).build());
  }

  @Override
  public String getName() {
    return "cons";
  }

  @Override
  public SCMCons apply2(Object arg1, Object arg2) {
    return cons(arg1, arg2);
  }

  public static SCMCons cons(Object car, Object cdr) {
    if (car == null && cdr == null) {
      return SCMCons.EMPTY;
    }
    return SCMCons.cons(car, cdr);
  }
}

File: src/main/java/core/procedures/cons/ConsProc.kt
package core.procedures.cons

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.procedures.predicates.Predicate
import core.utils.Utils

class ConsProc : AFn<Any?, Any?>(name = "cons", arity = Exactly(2)) {

    override operator fun invoke(arg1: Any?, arg2: Any?): Any = when (Predicate.isProperList(arg2)) {
        true  -> listOf(arg1) + Utils.toSequence(arg2)
        false -> Pair(arg1, arg2)
    }
}
